libraries/AP_Camera/AP_Camera_SoloGimbal.cpp

```diff
@@ -0,0 +1,110 @@
+    const uint8_t gopro_shutter_stop[4] = { 0, 0, 0, 0};
+
+    if (gopro_capture_mode == GOPRO_CAPTURE_MODE_PHOTO) {
+        // Trigger shutter start to take a photo
+        gcs().send_text(MAV_SEVERITY_INFO, "GoPro Photo Trigger");
+        mavlink_msg_gopro_set_request_send(heartbeat_channel, mavlink_system.sysid, MAV_COMP_ID_GIMBAL,GOPRO_COMMAND_SHUTTER,gopro_shutter_start);
+
+    } else if (gopro_capture_mode == GOPRO_CAPTURE_MODE_VIDEO) {
+        if (gopro_is_recording) {
+            // GoPro is recording, so stop recording
+            gcs().send_text(MAV_SEVERITY_INFO, "GoPro Recording Stop");
+            mavlink_msg_gopro_set_request_send(heartbeat_channel, mavlink_system.sysid, MAV_COMP_ID_GIMBAL,GOPRO_COMMAND_SHUTTER,gopro_shutter_stop);
+        } else {
+            // GoPro is not recording, so start recording
+            gcs().send_text(MAV_SEVERITY_INFO, "GoPro Recording Start");
+            mavlink_msg_gopro_set_request_send(heartbeat_channel, mavlink_system.sysid, MAV_COMP_ID_GIMBAL,GOPRO_COMMAND_SHUTTER,gopro_shutter_start);
+        }
+    } else {
+        gcs().send_text(MAV_SEVERITY_ERROR, "GoPro Unsupported Capture Mode");    
+    }
+}
+
+// Cycle the GoPro capture mode
+// This is so ArduPilot can toggle the capture mode of the GoPro directly, probably with an RC Aux function.
+// This is primarily for Solo's gimbal being installed on a vehicle other than a Solo.  The usual GoPro controls 
+// thorugh the Solo app and Solo controller do not use this, as it is done offboard on the companion computer.
+void AP_Camera_SoloGimbal::gopro_capture_mode_toggle()
+{
+    uint8_t gopro_capture_mode_values[4] = { };
+    
+    if (gopro_status != GOPRO_HEARTBEAT_STATUS_CONNECTED) {
+        gcs().send_text(MAV_SEVERITY_ERROR, "GoPro Not Available");
+        return;
+    }
+
+    if (gopro_capture_mode == GOPRO_CAPTURE_MODE_PHOTO) {
+        // Change to video mode
+        gopro_capture_mode_values[0] =  GOPRO_CAPTURE_MODE_VIDEO;
+        mavlink_msg_gopro_set_request_send(heartbeat_channel, mavlink_system.sysid, MAV_COMP_ID_GIMBAL,GOPRO_COMMAND_CAPTURE_MODE,gopro_capture_mode_values);
+        gcs().send_text(MAV_SEVERITY_INFO, "GoPro changing to mode video");
+    } else if (gopro_capture_mode == GOPRO_CAPTURE_MODE_VIDEO) {
+        if (gopro_is_recording) {
+            // GoPro is recording, cannot change modes
+            gcs().send_text(MAV_SEVERITY_INFO, "GoPro recording, can't change modes");
+        } else {
+            // Change to camera mode
+            gopro_capture_mode_values[0] = GOPRO_CAPTURE_MODE_PHOTO;
+            mavlink_msg_gopro_set_request_send(heartbeat_channel, mavlink_system.sysid, MAV_COMP_ID_GIMBAL,GOPRO_COMMAND_CAPTURE_MODE,gopro_capture_mode_values);
+            gcs().send_text(MAV_SEVERITY_INFO, "GoPro changing to mode photo");
+        }
+    } else {
+        gcs().send_text(MAV_SEVERITY_ERROR, "Unsupported GoPro Capture Mode");    
```

Comment:
Could warn when the bounds-checking code fails rather than when we go to use the value.

Reply:
I did then realized it would be sending the warning over and over at whatever hz the heartbeat is at.  For how unlikely it is and how few users there are...?